### PR TITLE
[WIP][Feature] Support Mask2former in MMSeg 1.x

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Support Mask2former.

Ref: 
1. [Official repo](https://github.com/facebookresearch/Mask2Former) 
2. [MMDetection](https://github.com/open-mmlab/mmdetection)

## Modification

1. add mask2former head
2. configs

## Current results:

|dataset|backbone|Ref mIoU|mIoU|
|---|---|---|---|
|Cityscapes|R50|79.4|80.16|

### CODE REVIEW

File: mmseg/models/decode_heads/mask2former_head.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+@MODELS.register_module()
+class Mask2FormerHead(Mask2FormerHead_):
+
+    def __init__(self,
+                 num_classes,
+                 align_corners=False,
+                 ignore_index=255,
+                 **kwargs):
+        super().__init__(**kwargs)
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/mask2former_head.py:24]: Perhaps we need to delete `super().__init__(**kwargs)` and completely rewrite `__init__` method?

File: configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+crop_size = (512, 1024)
+data_preprocessor = dict(
+    type='SegDataPreProcessor',
+    mean=[123.675, 116.28, 103.53],
+    std=[58.395, 57.12, 57.375],
+    bgr_to_rgb=True,
+    pad_val=0,
+    seg_pad_val=255,
+    size=crop_size)
+model = dict(
```
Comment (MengzhangLI) [inline configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py:14]: Add `num_class = 19` like [mmdet](https://github.com/open-mmlab/mmdetection/blob/3.x/configs/maskformer/maskformer_r50_ms-16xb1-75e_coco.py#L18).

File: configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                feedforward_channels=2048,
+                operation_order=('cross_attn', 'norm', 'self_attn', 'norm',
+                                 'ffn', 'norm')),
+            init_cfg=None),
+        loss_cls=dict(
+            type='CrossEntropyLoss',
+            use_sigmoid=False,
+            loss_weight=2.0,
+            reduction='mean',
+            class_weight=[1.0] * 19 + [0.1]),
```
Comment (MengzhangLI) [inline configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py:104]:
```suggestion
            class_weight=[1.0] * num_classes + [0.1]),
```

File: configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    train_cfg=dict(),
+    test_cfg=dict(mode='whole'))
+
+# optimizer
+optimizer = dict(
+    type='AdamW', lr=0.0001, betas=(0.9, 0.999), weight_decay=0.0005)
+optim_wrapper = dict(
+    type='OptimWrapper',
+    optimizer=optimizer,
+    clip_grad=None,
```
Comment (MeowZheng) [inline configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py:149]: please check clip_grad

https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/configs/ade20k/semantic-segmentation/Base-ADE20K-SemanticSegmentation.yaml#L29-L33

File: configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        type='PolyLR',
+        eta_min=0,
+        power=0.9,
+        begin=0,
+        end=90000,
+        by_epoch=False)
+]
+
+# training schedule for 90k
+train_cfg = dict(type='IterBasedTrainLoop', max_iters=90000, val_interval=9000)
```
Comment (MeowZheng) [inline configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py:165]: https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/configs/ade20k/semantic-segmentation/Base-ADE20K-SemanticSegmentation.yaml#L22

90k or 160k iters

Comment (xiexinch) [reply configs/mask2former/mask2former_r50_8xb2-90k_cityscapes-512x1024.py:165]: The official repo sets 90k for cityscapes and 160k for ADE20K.

https://github.com/facebookresearch/Mask2Former/blob/main/configs/cityscapes/semantic-segmentation/Base-Cityscapes-SemanticSegmentation.yaml#L22

https://github.com/facebookresearch/Mask2Former/blob/main/configs/ade20k/semantic-segmentation/Base-ADE20K-SemanticSegmentation.yaml#L22

**Pull request closed without merging**